zebra-rpc/src/methods/get_block_template_rpcs.rs

```diff
@@ -333,40 +333,113 @@ where
+            // Set up the loop.
+            let mut max_time_reached = false;
+
+            // The loop returns the server long poll ID,
+            // which should be different to the client long poll ID.
+            let (server_long_poll_id, chain_tip_and_local_time, mempool_txs) = loop {
+                // Check if we are synced to the tip.
+                // The result of this check can change during long polling.
+                //
+                // TODO:
+                // - add `async changed()` methods to ChainTip and ChainSyncStatus
+                //   (using `changed() -> Changed` and `impl Future<()> for Changed`)
+                check_synced_to_tip(network, latest_chain_tip.clone(), sync_status.clone())?;
+
+                // Fetch the state data and local time for the block template:
+                // - if the tip block hash changes, we must return from long polling,
+                // - if the local clock changes on testnet, we might return from long polling
+                //
+                // We always return after 90 minutes on mainnet, even if we have the same response,
+                // because the max time has been reached.
+                //
+                // TODO: timeout and exit the loop when max time is reached
+                let chain_tip_and_local_time =
+                    fetch_state_tip_and_local_time(state.clone()).await?;
 
-            // Fetch the mempool data for the block template:
-            // - if the mempool transactions change, we might return from long polling.
-            let mempool_txs = fetch_mempool_transactions(mempool).await?;
+                // Fetch the mempool data for the block template:
+                // - if the mempool transactions change, we might return from long polling.
+                //
+                // TODO:
+                // - add a `MempoolChange` type with an `async changed()` method.
+                // - if we are long polling, pause between state and mempool,
+                //   to allow transactions to re-verify (only works after PR #5841)
+                let mempool_txs = fetch_mempool_transactions(mempool.clone()).await?;
+
+                // - Long poll ID calculation
+                let server_long_poll_id = LongPollInput::new(
+                    chain_tip_and_local_time.tip_height,
+                    chain_tip_and_local_time.tip_hash,
+                    chain_tip_and_local_time.max_time,
+                    mempool_txs.iter().map(|tx| tx.transaction.id),
+                )
+                .generate_id();
+
+                // The loop finishes if:
+                // - the client didn't pass a long poll ID,
+                // - the server long poll ID is different to the client long poll ID, or
+                // - the previous loop iteration waited until the max time.
+                if Some(&server_long_poll_id) != client_long_poll_id.as_ref() || max_time_reached {
```

Comment:
I think this would return a template with a `max_time` in the past if `max_time_reached` is true but the long poll id is the same.

Reply:
Consensus rules can't depend on the local node time, because they need to be global. So it's definitely unusual, but I think it's still valid?

`max_time` is clamped to 90 minutes later than the `median-time-past` of recent blocks, based on this consensus rule:
https://github.com/ZcashFoundation/zebra/blob/4fb2417adc70f57f42e9df4c8467279058b743d7/zebra-state/src/service/read/difficulty.rs#L213-L219

(There is a standard 2 hour rule for the local node time, but standard rules can vary between implementations or configurations, so we don't use it to construct the block template.)